client-sdk/go/crypto/signature/context.go

```diff
@@ -1,27 +1,68 @@
 package signature
 
 import (
+	"encoding/hex"
+
 	"github.com/oasisprotocol/oasis-core/go/common"
 	"github.com/oasisprotocol/oasis-core/go/common/crypto/hash"
+
+	ethCommon "github.com/ethereum/go-ethereum/common"
 )
 
 const chainContextSeparator = " for chain "
 
-// Context is the chain domain separation context.
-type Context string
+// SignContext stores runtime ID, consensus chain context and tx-specific details
+// as defined in ADR 14 appropriate for CBOR encoding.
+type SignContext struct {
+	base         []byte
+	RuntimeID    string `json:"runtime_id"`
+	ChainContext string `json:"chain_context"`
+	OrigTo       string `json:"orig_to,omitempty"`
+}
+
+// Derive derives the chain domain separation context for a given runtime and consensus chain.
+func (sc SignContext) Derive() []byte {
+	// Ignore context derivation, if Runtime ID and consensus chain context are not set.
+	if sc.RuntimeID == "" && sc.ChainContext == "" {
+		return []byte{}
+	}
 
-func (c Context) New(base []byte) []byte {
-	ctx := append([]byte{}, base...)
+	ctx := append([]byte{}, sc.base...)
 	ctx = append(ctx, []byte(chainContextSeparator)...)
+
+	rawRuntimeID, _ := hex.DecodeString(sc.RuntimeID)
+	c := hash.NewFromBytes(
+		rawRuntimeID,
+		[]byte(sc.ChainContext),
+	).String()
 	ctx = append(ctx, []byte(c)...)
+
 	return ctx
 }
 
-// DeriveChainContext derives the chain domain separation context for a given runtime.
-func DeriveChainContext(runtimeID common.Namespace, consensusChainContext string) Context {
-	rawRuntimeID, _ := runtimeID.MarshalBinary()
-	return Context(hash.NewFromBytes(
-		rawRuntimeID,
-		[]byte(consensusChainContext),
-	).String())
+// AddTxDetails adds optional transaction details to the signature context.
+func (sc *SignContext) AddTxDetails(txDetails TxDetails) error {
```

Comment:
This seems like a convenience method and type to transfer some arguments (e.g. `OrigTo`) to populate SignContext. Maybe it would make more sense to define it in the caller (e.g. in the CLI) and not here as the use case outside the CLI is not that obvious.